kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -1122,6 +1143,26 @@ protected void impersonateUser(String userId) throws SakaiException {
 		sakaiSession.setUserId(validatedUserId);
 		sakaiSession.setUserEid(validatedUserEid);
 		authzGroupService().refreshUser(validatedUserId);
+		log.info("Enter into Role-Swap mode, user " + validatedUserId + " is impersonated");
 	}
-	
+
+	protected void restoreUser() throws SakaiException {
+		if (originalUserId == null || originalUserEid == null) {
+			throw new SakaiException("Original user session information is missing");
+		}
+
+		Session sakaiSession = sessionManager().getCurrentSession();
+
+		// Restore the original user session
+		Vector saveAttributes = new Vector();
```

Comment:
Unless you need Vector's built in synchronisation you should probably use a Collection here
Collection<String> saveAttributes = new ArrayList<>();

Reply:
Ah, right. It has to be a vector ..